File: src/fmj/views/ScreenAnimation.kt
package fmj.views

import fmj.Global
import fmj.ScreenViewType
import fmj.lib.DatLib
import fmj.lib.ResSrs
import graphics.Canvas

class ScreenAnimation(override val parent: GameNode, private val index: Int) : BaseScreen {
    override val screenName: String = "ScreenAnimation($index)"
    private val mResSrs: ResSrs

    init {
        if (index != 247 && index != 248 && index != 249) {
            throw IllegalArgumentException("只能是247,248,249")
        }
        mResSrs = DatLib.getRes(DatLib.ResType.SRS, 1, index) as ResSrs
        mResSrs.setIteratorNum(4)
        mResSrs.start()
    }

    override fun update(delta: Long) {
        if (!mResSrs.update(delta)) {
            when (index) {
                247 -> // 转到游戏动画
                    game.changeScreen(ScreenViewType.SCREEN_GAME_LOGO)
                248 -> // 转到游戏菜单
                    game.changeScreen(ScreenViewType.SCREEN_MENU)
                249 -> //
                    game.changeScreen(ScreenViewType.SCREEN_MENU)
            }
        }
    }

    override fun draw(canvas: Canvas) {
        canvas.drawColor(Global.COLOR_WHITE)
        // 假设动画原始尺寸为160x96，在320x192屏幕中居中显示
        val animWidth = 160
        val animHeight = 96
        val centerX = (Global.SCREEN_WIDTH - animWidth) / 2
        val centerY = (Global.SCREEN_HEIGHT - animHeight) / 2
        mResSrs.draw(canvas, centerX, centerY)
    }

    override fun onKeyDown(key: Int) {
        if (key == Global.KEY_CANCEL && (index == 247 || index == 248)) {
            game.changeScreen(ScreenViewType.SCREEN_MENU)
        }
    }

    override fun onKeyUp(key: Int) {}

}

File: packages/core/src/views/screen-animation.ts
import type { Game } from '@/game/game';
import { ResSrs } from '@/lib/res-srs';
import { ResourceType } from '@/lib/resource-utils';
import { Surface } from '@/rendering/surface';
import { COLOR_WHITE } from '@/rendering/color';
import { SCREEN_HEIGHT, SCREEN_WIDTH } from '@/shared/constants';
import { KeyCode } from '@/shared/key-code';
import { BaseScreen } from './base-screen';
import { ScreenViewType } from './screen-view-type';

const STARTUP_WIDTH = 160;
const STARTUP_HEIGHT = 96;

export class ScreenAnimation extends BaseScreen {
  private readonly animation: ResSrs;
  private readonly index: number;

  constructor(game: Game, index: number) {
    super(game);
    this.index = index;
    if (index !== 247 && index !== 248 && index !== 249) {
      throw new Error('ScreenAnimation index must be 247, 248, or 249');
    }

    const resource = this.game.datLib.getRes(ResourceType.SRS, 1, index);
    if (!(resource instanceof ResSrs)) {
      throw new Error(`Missing SRS animation 1:${index}`);
    }

    this.animation = resource;
    this.animation.setIteratorNum(4);
    this.animation.start();
  }

  override update(delta: number): void {
    if (this.animation.update(delta)) return;

    switch (this.index) {
      case 247:
        this.game.changeScreen(ScreenViewType.SCREEN_GAME_LOGO);
        return;
      case 248:
      case 249:
        this.game.changeScreen(ScreenViewType.SCREEN_MENU);
        return;
    }
  }

  draw(surface: Surface): void {
    surface.drawColor(COLOR_WHITE);
    const centerX = Math.floor((SCREEN_WIDTH - STARTUP_WIDTH) / 2);
    const centerY = Math.floor((SCREEN_HEIGHT - STARTUP_HEIGHT) / 2);
    this.animation.draw(surface, centerX, centerY);
  }

  override onKeyDown(key: KeyCode): void {
    if (key === KeyCode.Cancel && (this.index === 247 || this.index === 248)) {
      this.game.changeScreen(ScreenViewType.SCREEN_MENU);
    }
  }
}
